File: csharp/step0_repl.cs
using System;

namespace Mal
{
    internal class MAL
    {
        static int Main(string[] args)
        {
            return Repl();
        }

        static string READ(string arg) => arg;

        static string EVAL(string arg) => arg;

        static string PRINT(string arg) => arg;

        static string Rep(string arg)
        {
            return PRINT(EVAL(READ (arg)));
        }

        internal static int Repl()
        {
            for (; ;)
            {
                Console.Write("user> ");
                var str = Console.ReadLine();
                if (str == null)
                    break;
                Console.WriteLine(Rep(str));
            }

            return 0;
        }
    }
}

File: cxx/step0_repl.cxx
#include <iostream>
#include <string>
#include <algorithm>

#include <readline/readline.h>
#include <readline/history.h>

using String = std::string;

String READ(const String arg)
{
   return arg;
}

String EVAL(const String arg)
{
   return arg;
}

String PRINT(const String arg)
{
   return arg;
}

String rep(const String arg)
{
   return PRINT(EVAL(READ(arg)));
}

bool readline(const char * prompt, String & output)
{
   char * line = readline(prompt);
   if (line == nullptr)
      return false;

   output = line;

   auto pos = std::find_if(output.cbegin(), output.cend(),
                           [](const char ch) { return (ch > ' '); });

   if (pos != output.cend())
      add_history(line);

   free(line);

   return true;
}

int main(int argc, char * argv[])
{
   String input;

   while (readline("user> ", input))
   {
      std::cout << rep(input) << std::endl;
   }
   return 0;
}
